packages/block-editor/src/hooks/block-binding-support/with-block-binding-support.js

```diff
@@ -0,0 +1,178 @@
+import { getBlockType } from '@wordpress/blocks';
+
+/**
+ * This component is responsible detecting and
+ * propagating data changes between block attribute and
+ * the block-binding source property.
+ *
+ * The app creates an instance of this component for each
+ * pair of block-attribute/source-property.
+ *
+ * @param {Object}   props            - The component props.
+ * @param {string}   props.attrName   - The attribute name.
+ * @param {any}      props.attrValue  - The attribute value.
+ * @param {Function} props.useSource  - The custom hook to use the source.
+ * @param {Object}   props.blockProps - The block props with bound attribute.
+ * @param {Object}   props.args       - The arguments to pass to the source.
+ * @return {null}                       This is a data-handling component. Render nothing.
+ */
+const BlockBindingConnector = ( {
+	args,
+	attrName,
+	attrValue,
+	blockProps,
+	useSource,
+} ) => {
+	const {
+		placeholder,
+		value: propValue,
+		updateValue: updatePropValue,
+	} = useSource( blockProps, args );
+
+	const blockName = blockProps.name;
+
+	const setAttributes = blockProps.setAttributes;
+
+	const updateBoundAttibute = useCallback(
+		( newAttrValue ) => {
+			setAttributes( {
+				[ attrName ]: newAttrValue,
+			} );
+		},
+		[ attrName, setAttributes ]
+	);
+
+	// Store a reference to the last value and attribute value.
+	const lastPropValue = useRef( propValue );
+	const lastAttrValue = useRef( attrValue );
+
+	/*
+	 * Initially sync (first render / onMount ) attribute
+	 * value with the source prop value.
+	 */
+	useEffect( () => {
+		updateBoundAttibute( propValue );
+	}, [ propValue, updateBoundAttibute ] );
+
+	/*
+	 * Sync data.
+	 * This effect will run every time
+	 * the attribute value or the prop value changes.
+	 * It will sync them in both directions.
+	 */
+	useEffect( () => {
+		/*
+		 * Source Prop => Block Attribute
+		 *
+		 * Detect changes in source prop value,
+		 * and update the attribute value accordingly.
+		 */
+		if ( typeof propValue !== 'undefined' ) {
+			if ( propValue !== lastPropValue.current ) {
+				lastPropValue.current = propValue;
+				updateBoundAttibute( propValue );
+				return;
+			}
+		} else if ( placeholder ) {
+			/*
+			 * If the attribute is `src` or `href`,
+			 * a placeholder can't be used because it is not a valid url.
+			 * Adding this workaround until
+			 * attributes and metadata fields types are improved and include `url`.
+			 */
+			const htmlAttribute =
+				getBlockType( blockName ).attributes[ attrName ].attribute;
+
+			if ( htmlAttribute === 'src' || htmlAttribute === 'href' ) {
+				updateBoundAttibute( null );
+				return;
+			}
+
+			updateBoundAttibute( placeholder );
+		}
+
+		/*
+		 * Block Attribute => Source Prop
+		 *
+		 * Detect changes in block attribute value,
+		 * and update the source prop value accordingly.
+		 */
+		if ( attrValue !== lastAttrValue.current && updatePropValue ) {
```

Comment:
Same comment as in https://github.com/WordPress/gutenberg/pull/58895/files#r1490121529

> We should check here if the `propValue` or the `lastPropValue.current` is an instance of `RichTextData` because it might be the case for Paragraph blocks.
>
> If so, we'd have to get the value from RichTextData instance with `.text()` and compare those values.